pkg/kernelmonitor/logwatchers/journald/log_watcher.go

```diff
@@ -0,0 +1,163 @@
+type journaldWatcher struct {
+	journal *sdjournal.Journal
+	cfg     types.WatcherConfig
+	logCh   chan *kerntypes.KernelLog
+	tomb    *util.Tomb
+}
+
+// NewJournaldWatcher is the create function of journald watcher.
+func NewJournaldWatcher(cfg types.WatcherConfig) types.LogWatcher {
+	return &journaldWatcher{
+		cfg:  cfg,
+		tomb: util.NewTomb(),
+		// A capacity 1000 buffer should be enough
+		logCh: make(chan *kerntypes.KernelLog, 1000),
+	}
+}
+
+// Make sure NewJournaldWatcher is types.WatcherCreateFunc .
+var _ types.WatcherCreateFunc = NewJournaldWatcher
+
+// Watch starts the journal watcher.
+func (j *journaldWatcher) Watch() (<-chan *kerntypes.KernelLog, error) {
+	journal, err := getJournal(j.cfg)
+	if err != nil {
+		return nil, err
+	}
+	j.journal = journal
+	glog.Info("Start watching journald")
+	go j.watchLoop()
+	return j.logCh, nil
+}
+
+// Stop stops the journald watcher.
+func (j *journaldWatcher) Stop() {
+	j.tomb.Stop()
+}
+
+// waitLogTimeout is the timeout waiting for new log.
+const waitLogTimeout = 5 * time.Second
+
+// watchLoop is the main watch loop of journald watcher.
+func (j *journaldWatcher) watchLoop() {
+	defer func() {
+		if err := j.journal.Close(); err != nil {
+			glog.Errorf("Failed to close journal client: %v", err)
+		}
+		j.tomb.Done()
+	}()
+	for {
+		select {
+		case <-j.tomb.Stopping():
+			glog.Infof("Stop watching journald")
+			return
+		default:
+		}
+		// Get next log entry.
+		n, err := j.journal.Next()
+		if err != nil {
+			glog.Errorf("Failed to get next journal entry: %v", err)
+			continue
+		}
+		// If next reaches the end, wait for waitLogTimeout.
+		if n == 0 {
+			j.journal.Wait(waitLogTimeout)
+			continue
+		}
+
+		entry, err := j.journal.GetEntry()
+		if err != nil {
+			glog.Errorf("failed to get journal entry: %v", err)
+			continue
+		}
+
+		j.logCh <- translate(entry)
+	}
+}
+
+// defaultJournalLogPath is the default path of journal log.
+const defaultJournalLogPath = "/var/log/journal"
+
+// getJournal returns a journal client.
+func getJournal(cfg types.WatcherConfig) (*sdjournal.Journal, error) {
+	// Get journal log path.
+	path := defaultJournalLogPath
+	if cfg.LogPath != "" {
+		path = cfg.LogPath
+	}
+	// Get lookback duration.
+	since, err := util.ParseDuration(cfg.Lookback)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse lookback duration %q: %v", cfg.Lookback, err)
+	}
+	// Get journal client from the log path.
+	journal, err := sdjournal.NewJournalFromDir(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create journal client from path %q: %v", path, err)
+	}
+	// Seek journal client based on the lookback duration.
+	start := time.Now().Add(-since)
+	err = journal.SeekRealtimeUsec(uint64(start.UnixNano() / 1000))
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookback %q: %v", since, err)
+	}
+	// TODO(random-liu): Make this configurable to support parsing other logs.
+	kernelMatch := sdjournal.Match{
```

Comment:
Some people may want to use this to monitor other logs (e.g. docker log) to catch some known docker issue. (See https://github.com/kubernetes/node-problem-detector/issues/44_